From 682c5f6a0a43f0766badbae9633f0855d7a5acef Mon Sep 17 00:00:00 2001
From: Chunwei Chen <tuxoko@gmail.com>
Date: Sat, 22 Feb 2025 10:50:33 -0800
Subject: [PATCH 35/53] Fix wrong free function in arc_hdr_decrypt

Need to use arc_free_data_abd to free abd type buffer.

Reviewed-by: Alexander Motin <mav@FreeBSD.org>
Signed-off-by: Chunwei Chen <tuxoko@gmail.com>
Closes #17079
---
 module/zfs/arc.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/module/zfs/arc.c b/module/zfs/arc.c
index 1f653d953..46f0582ba 100644
--- a/module/zfs/arc.c
+++ b/module/zfs/arc.c
@@ -1905,7 +1905,7 @@ arc_hdr_decrypt(arc_buf_hdr_t *hdr, spa_t *spa, const zbookmark_phys_t *zb)
 error:
 	arc_hdr_free_abd(hdr, B_FALSE);
 	if (cabd != NULL)
-		arc_free_data_buf(hdr, cabd, arc_hdr_size(hdr), hdr);
+		arc_free_data_abd(hdr, cabd, arc_hdr_size(hdr), hdr);
 
 	return (ret);
 }
-- 
2.39.5

